lib/spack/spack/solver/asp.py

```diff
@@ -115,8 +115,30 @@ def getter(node):
     "VersionProvenance", version_origin_fields
 )(**{name: i for i, name in enumerate(version_origin_fields)})
 
-#: Named tuple to contain information on declared versions
-DeclaredVersion = collections.namedtuple("DeclaredVersion", ["version", "idx", "origin"])
+
+class DeclaredVersion(object):
+    def __init__(self, version: spack.version.VersionBase, idx, origin):
+        if not isinstance(version, spack.version.VersionBase):
+            raise ValueError("Unexpected type for declared version: {0}".format(type(version)))
+        self.version = version
+        self.idx = idx
+        self.origin = origin
+
+    def __eq__(self, other):
+        if not isinstance(other, DeclaredVersion):
+            return False
+        is_git = lambda v: isinstance(v, spack.version.GitVersion)
+        if is_git(self.version) != is_git(other.version):
+            # GitVersion(hash=x) and Version(y) compare equal if x == y, but
+            # we do not want that to be true for DeclaredVersion, which is
+            # tracking how much information is provided: in that sense, the
+            # GitVersion provides more information and is therefore not equal.
+            return False
```

Comment:
@scheibelp: This should just ensure that all the CLI DeclaredVersions have a higher precedence than the others.  I think you needed the complex DeclaredVersion class above because you introduced a potentially duplicate `idx` here.

In general it's good for the weights in the concretizer (which the `idx` fields end up being used for) to be distinct, anyway, as it makes for deterministic answers.

Reply:
@tgamblin this concern is ostensibly resolved in https://github.com/spack/spack/pull/36347/commits/c92938fe6afa6725388dd3e4c4234b793602675b, but see https://github.com/spack/spack/pull/36347#issuecomment-1481973196 for a more-careful treatment of the related concerns